velox/vector/DecodedVector.h

```diff
@@ -158,6 +158,18 @@ class DecodedVector {
     return reinterpret_cast<const T*>(data_);
   }
 
+  /// Returns the value at the specified index in the values buffer.
+  /// Note: This method doesn't use the mapping in indices() to resolve the
+  /// index. Use this method if DecodedVector::isIdentityMapping() returns true.
+  template <typename T>
+  const T deserialize(const vector_size_t idx) const {
```

Comment:
This API is inconsistent with the reset of the class and looks suspicious. Can we find a way to remove it?

Also, I do not understand what problem we are trying to solve? Aren't we already allocating memory for aggregate accumulators with the proper alignment? Would you please explain?

Reply:
Initializing as well as reading from character buffer has been a known issue with int128_t on linux systems. Earlier, I had specialized reading int128_t in the Aggregate class itself. As per @Yuhta suggestion here:https://github.com/facebookincubator/velox/pull/3755#discussion_r1117628181, I had to abstract it out in the DecodedVector class.

Reply:
Can we dig it more into the cause?  Because I am not aware of any other issue then alignment that can cause issue reading `int128_t`.  Do we have a minimal example to reproduce the issue?

Reply:
I have reproduced it several times on centos docker container. If we undo the deserialize changes in this PR and run circleci wf you must be able to reproduce it. This is not the only location where we had issue reading int128_t values. Refer: https://github.com/prestodb/presto/pull/18838/files

Reply:
Do we know if it is due to alignment or something else?  If it is something else, do we have a bug to track in the compiler?

Reply:
> Do we have a minimal example to reproduce the issue?

@karteekmurthys Karteek, it would be super helpful to create a small unit test that demonstrates the issue (without using any of the aggregate function interfaces.)

Reply:
https://github.com/prestodb/presto/pull/18838/files has some functional tests without using agg interface. It is reproducible only on linux machines and release builds.

Reply:
Perhaps, a cleaner fix is to provide a template override for valueAt for decimal type.

Reply:
BTW, does it mean that FlatVector<T>::valueAt also doesn't work for decimals?

Reply:
We are already overriding valueAt for long decimals. `deserialize` method is required for isIdentityMapping() case where we use values buffer itself to get the data and avoid valueAt<>().

```
template <>
inline UnscaledLongDecimal DecodedVector::valueAt(vector_size_t idx) const {
  auto valuePosition = reinterpret_cast<const char*>(data_) +
      sizeof(UnscaledLongDecimal) * index(idx);
  return UnscaledLongDecimal::deserialize(valuePosition);
}
```

Reply:
FlatVector::valueAt -> valueATFast

```
template <typename T>
const T FlatVector<T>::valueAtFast(vector_size_t idx) const {
  return rawValues_[idx];
}
```

There is not reinterpret_cast here. Using cast on char buffer to int128_t is where we see the issue.

Reply:
Then let's treat decimal the same as bool and use valueAt for it.

Reply:
But, this means we will lose the optimization right?

Reply:
deserialize logic is already not cheap, hence, may not matter.

Reply:
Done.